### ISSUE TITLE
Adasum Algorithm leads to convergence degradation

### ISSUE THREAD

**Issue by PANXiao1994** [horovod/horovod#1667]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 1.15.0
3. Horovod version: unreleased version by using `HOROVOD_NCCL_HOME=/usr HOROVOD_GPU_BROADCAST=NCCL HOROVOD_WITH_TENSORFLOW=1 HOROVOD_GPU_ALLREDUCE=NCCL pip3 install git+https://github.com/horovod/horovod.git`
4. MPI version: Open MPI: 2.0.2
5. CUDA version: CUDA Version 10.0.130
6. NCCL version: libnccl-dev=2.4.8-1
7. Python version: python 3.5
8. OS and version: Debian GNU/Linux,   9 (stretch)
9. GCC version: gcc version 6.3.0 20170516 (Debian 6.3.0-18+deb9u1)

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
yes. but no reply
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Bug report:**
also see: https://github.com/horovod/horovod/issues/1653

I did a set of comparative experiments, where both tasks have hvs.size=8  and same batch_size / gpu.
The only difference is the `op` parameter in DistributedOptimizer:

For task 1:
```
optimizer = hvd.DistributedOptimizer(self._optimizer, sparse_as_dense=True, backward_passes_per_step=self.params["update_cycle"], op=hvd.mpi_ops.Adasum)
```

and for task 2:
```
optimizer = hvd.DistributedOptimizer(self._optimizer, sparse_as_dense=True, backward_passes_per_step=self.update_cycle, op=hvd.mpi_ops.Average)
```

p.s. Note that here update_cycle = 1.

This is the loss convergence curves:
![image](https://user-images.githubusercontent.com/24204710/72318421-2da6dc00-36d7-11ea-8593-51bfbf257e5f.png)
where the red one corresponds to task 1 (Adasum) and the purple one corresponds to task 2 (Average).

We can observe that by turning `Average` into `Adasum`, loss convergence suffers a lot. Could anybody help this? @Tixxx @tgaddair

**Comment by tgaddair** [comment]
Hey @PANXiao1994, how are you scaling your learning rate?  With Adasum + NCCL, it should be `lr * hvd.local_size()` as opposed to the normal `lr * hvd.size()`.

**Comment by Tixxx** [comment]
Thanks @PANXiao1994  for doing the experiments. As Travis already suggested, Adasum does require a bit of change when migrating from sum/average. Generally we will tell users to start with ` lr * hvd.local_size()` for Adasum + NCCL and just `lr` for only Adasum. We are working on an user guideline and will publish it soon.

**Comment by PANXiao1994** [comment]
> Thanks @PANXiao1994 for doing the experiments. As Travis already suggested, Adasum does require a bit of change when migrating from sum/average. Generally we will tell users to start with ` lr * hvd.local_size()` for Adasum + NCCL and just `lr` for only Adasum. We are working on an user guideline and will publish it soon.

I used Adam Optimizer with noam decay
![image](https://user-images.githubusercontent.com/24204710/72399417-16242d80-3781-11ea-992c-e403b9f0e9a6.png)

So what I need to do is to set `scale *= hvd.local_size()`?

**Comment by tgaddair** [comment]
Hey @PANXiao1994, in your experiments both setups converged in the same number of steps, but was there any difference in the time taken for each step to complete (alternatively, total time to convergence)?

**Comment by PANXiao1994** [comment]
> Hey @PANXiao1994, in your experiments both setups converged in the same number of steps, but was there any difference in the time taken for each step to complete (alternatively, total time to convergence)?

@tgaddair  No, there is no difference. In experience 1, where I reduced hvd.local_size from 8 to 1, the time for each step keeps the same...

**Comment by PANXiao1994** [comment]
@tgaddair I am pretty sure now that the time for each step to finish when I use horovod with Adasum keeps exactly the same as I train with such batch size / gpu on a single GPU... ...

**Comment by rosrad** [comment]
hi @tgaddair and @PANXiao1994

I have observed the similar trend that the adasum do not convert to a better point
I have done a comparison experiment with 3 differernt setting
with avg, adasum, adasum with lr*hvd.local_sie
the experiment are trained in the same 1x4 gpus
I found that

with the same lr, the adasum do not convert to a better point
enlarge the lr by hvd.local_size(), the adasum do not help, but become worse
![image](https://user-images.githubusercontent.com/2343640/72860549-1af95c00-3d02-11ea-8d76-80c502f74a01.png)

**Comment by Tixxx** [comment]
Hi @PANXiao1994, trying to understand your experiments better. In your experiment 1, are you saying the total time to convergence is the same between local_size=8 and local_size=1?
For your experiment 2 where you use 8 gradient accumulations for Adasum, have you measured total time to convergence too? With 8 local accumulations Adasum is essentially doing way less communications than Averaging so total time should be less.

**Comment by Tixxx** [comment]
Hi @rosrad Was your experiment run with Horovod compiled with `HOROVOD_GPU_ALLREDUCE=NCCL`?

**Comment by abditag2** [comment]
> > Thanks @PANXiao1994 for doing the experiments. As Travis already suggested, Adasum does require a bit of change when migrating from sum/average. Generally we will tell users to start with ` lr * hvd.local_size()` for Adasum + NCCL and just `lr` for only Adasum. We are working on an user guideline and will publish it soon.
> 
> My experiments is conducted on Machine Translation using Transformer architecture.
> 
> I used Adam Optimizer with noam decay
> ![image](https://user-images.githubusercontent.com/24204710/72399417-16242d80-3781-11ea-992c-e403b9f0e9a6.png)
> 
> where `warmup_steps = decay_step * decay_step_scale`
> 
> * `decay_step`: 4000
> * `decay_step_scale`: max(min(25000. / true_batch_tokens, 1.), 0.25)
>   
>   * `true_batch_tokens`: batch tokens size / gpu * hvd.size()
> 
> So `warmup_steps` depends on batch tokens size. This is why I don't know how to modify the learning rate.

If your batch size per GPU is `bs1, to set your decay_step_scale, you can use `bs*num_gpus` as your batch size for average and `bs*num_local_gpus` in the Adasum case (assuming all your workers have same local size). This will make sure you have compatible learning rates.
